Provide submission summary when tapping a submission on the form map

grzesiek2010 opened the pull request:
Closes #3737

#### What has been done to verify that this works as intended?
I tested implemented changes manually.

#### Why is this the best possible solution? Were any other approaches considered?
This is what we discussed on the forum https://forum.getodk.org/t/provide-submission-summary-when-tapping-a-submission-on-the-form-map/25957

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
It just changes the way we deal with submissions visible on a map. Only this functionality might be affected. Now we should see a submission every time we click on a marked instead of opening a form immediately.

#### Do we need any specific form for testing your changes? If so, please attach one.
Any form with geopoints.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [ ] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [ ] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [ ] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

grzesiek2010 committed:
Added submission_summary_layout

grzesiek2010 committed:
Improved instance_map_layout

grzesiek2010 committed:
Improved submission_summary_layout

grzesiek2010 committed:
Added required icons

grzesiek2010 committed:
Show bottom sheet after clicking on a marker

grzesiek2010 committed:
Center the selected marker

grzesiek2010 committed:
Removed unused import

grzesiek2010 committed:
Display date and status in one row like we do in lists of instances

grzesiek2010 commented:
@lognaturel this is the pr you wanted for review so here you go.

lognaturel commented:
I think that this can move forward with a more final implementation and feature tests. I responded to the forum thread about the action button. I’d rather use another visual pattern for that. That said, I’m ok with you and @seadowg discussing and deciding which way to go.

grzesiek2010 committed:
Replaced FAB with a Chip

grzesiek2010 committed:
Fixed bug - open read_only form only for review

grzesiek2010 commented:
@lognaturel I pushed the latest changes in case you want to play with it on your device. I saw :heart: on the forum so I guess you like it more than the version with FAB but I'll wait for your feedback.

lognaturel commented:
This looks fantastic to me. The action is explicit and it's where the eye is already to get summary info.

I had completely missed this tag -- as we're getting close to code freeze or things get time-sensitive in some other way, please consider also asking for review in #collect-code so there's less risk of an accidental miss (sorry).

grzesiek2010 committed:
Added missing test props

grzesiek2010 committed:
Fixed naming bug

grzesiek2010 committed:
Improved submission_summary_layout

grzesiek2010 committed:
Refactored BottomSheet to BottomSheetDialog

grzesiek2010 committed:
Fixed tests

grzesiek2010 committed:
Fixed code style

grzesiek2010 committed:
Reverted redundant changes

lognaturel commented:
I'm happy with the functionality and user experience. I expected the bottom sheet to be gone when I came back to the map after viewing/editing a submission. I can see the logic either way and it's a non-disruptive thing to change so I think we can leave it as-is for now.

I'd like to have seen the logic in the ViewModel. Moving forward, we should probably favor the [Saved State ViewModel module](https://developer.android.com/topic/libraries/architecture/viewmodel-savedstate) when UI state needs to be maintained when the activity is stopped. We hadn't discussed this before so maybe it's best to leave this as-is and make the switch moving forward.

Test-wise, I think it would be nice to have a test specifically focused on the appearance of the bottom sheet when a submission is tapped.

@seadowg I didn't have time for a full review today and I'm not very confident in my two comments above. Could you please have a look?

grzesiek2010 committed:
Added displeying the summary dialog for encrypted and deleted forms ans well to keep consistency

grzesiek2010 committed:
Dismiss dialog after opening a form

grzesiek2010 committed:
Code improvements

grzesiek2010 committed:
Code improvements in order to use FormMapViewModel in InstanceSummaryDialogFragment

grzesiek2010 committed:
Fixed comment

grzesiek2010 committed:
Factored out getSubmissionSummaryStatusIcon to IconUtils

grzesiek2010 committed:
Code improvements

grzesiek2010 committed:
Use showIfNotShowing to display InstanceSummaryDialogFragment

grzesiek2010 committed:
Fixed status icon size

grzesiek2010 commented:
I noticed that if a submission is deleted or encrypted we clear coordinates:
https://github.com/getodk/collect/blob/master/collect_app/src/main/java/org/odk/collect/android/provider/InstanceProvider.java#L312
https://github.com/getodk/collect/blob/master/collect_app/src/main/java/org/odk/collect/android/tasks/SaveFormToDisk.java#L462

but we still have code for displaying info about them in FormMapActivity and test that are failing in this pr.
So I wonder whether it make sense to fix them maybe I can just remove the code and tests if it's not possible to display such instances on a map at all.
@lognaturel

seadowg commented on collect_app/src/main/res/drawable/form_state_finalized.xml:
```diff
@@ -1,20 +1,17 @@
-            android:pathData="M60,60m-60,0a60,60 0,1 1,120 0a60,60 0,1 1,-120 0"
-            android:fillColor="@color/deep_purple_500"/>
+        android:pathData="M67.195,26.83h-26.83c-3.688,0 -6.673,2.985 -6.673,6.633L33.66,86.537c0,3.648 2.984,6.634 6.673,6.634L80.61,93.171c3.689,0 6.707,-2.986 6.707,-6.634L87.317,46.732L67.195,26.829zM40.365,86.537L40.365,33.463h23.476L63.841,50.05L80.61,50.05v36.488L40.366,86.538z"
+        android:fillColor="?iconColor"/>
```
I think it's best avoid `iconColor` and go for Material Theme attributes like `colorOnSurface` or `colorOnPrimary`. These let us account for the fact that in some color schemes the icon color would differ depending on where it was rendered.

grzesiek2010 replied:
That's right but `?iconColor` is used in many places like that so would be better to clear that in a separate pr.

seadowg replied:
Absolutely. It's one of the follow ups to #3777 I'll schedule. It would be nice to not have another one to change in that though!

seadowg commented on collect_app/src/main/res/layout/submission_summary_layout.xml:
```diff
@@ -0,0 +1,70 @@
+
+        <TextView
+            android:id="@+id/submission_name"
+            style="@style/TextAppearance.AppCompat.Headline"
```
This should use of our Material Theme text appearances (`TextAppearance.Collect.X`). I'm guessing `Headline6` is too big but have a play with it!

grzesiek2010 replied:
Headline6 looks pretty good. Done.

seadowg commented on collect_app/src/main/res/layout/submission_summary_layout.xml:
```diff
@@ -0,0 +1,70 @@
+            style="@style/TextAppearance.AppCompat.Headline"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:textColor="?primaryTextColor" />
```
This should be `colorOnSurface`. `primaryTextColor` is actually gone on master.

grzesiek2010 replied:
Done.

seadowg commented on collect_app/src/main/res/layout/submission_summary_layout.xml:
```diff
@@ -0,0 +1,70 @@
+            android:layout_height="30dp" />
+
+        <TextView
+            android:id="@+id/status_text"
```
Could we use a text appearance for this as well?

grzesiek2010 replied:
Done.

seadowg commented on collect_app/src/main/res/layout/submission_summary_layout.xml:
```diff
@@ -0,0 +1,70 @@
+        app:chipEndPadding="@dimen/margin_small"/>
+
+    <TextView
+        android:id="@+id/info"
```
Again using a standard text appearance would be nice!

grzesiek2010 replied:
Done.

seadowg reviewed (changes requested):
This is a cool looking feature!

> Test-wise, I think it would be nice to have a test specifically focused on the appearance of the bottom sheet when a submission is tapped.

Yup I agree with that. Also left some comment inline mostly around theme attributes. Also had some questions around some of the architecture but I've reached out on Slack so we can chat.

grzesiek2010 committed:
Code improvements - use just instanceId in InstanceSummaryDialogFragment to build it

grzesiek2010 committed:
Layout improvements

lognaturel commented:
> maybe I can just remove the code and tests if it's not possible to display such instances on a map at all

I think that’s ok. They’re certainly not expected or desired states.

grzesiek2010 committed:
Fixed tests

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/FormMapActivity.java:
```diff
@@ -223,47 +216,10 @@ public void onLocationChanged(MapPoint point) {
-        Intent intent = getEditFormInstanceIntentFor(featureId);
-        intent.putExtra(ApplicationConstants.BundleKeys.FORM_MODE, ApplicationConstants.FormModes.VIEW_SENT);
-        return intent;
+        showSummary(featureId);
```
It doesn't make sense to have both `onFeatureClicked` and `showSummary` since they have the same parameters and `showSummary` is only called from here. I'd keep `onFeatureClicked` and move the code here but you could also rename it to `showSummary` if you think that's clearer.

grzesiek2010 replied:
Fixed.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/fragments/dialogs/InstanceSummaryDialogFragment.java:
```diff
@@ -0,0 +1,144 @@
+
+    @Override
+    public void onSaveInstanceState(Bundle outState) {
+        outState.putLong(INSTANCE_ID, instanceId);
```
This is much better and I guess that if truly only necessary data is being saved continuing to use `onSaveInstanceState` is not so bad.

grzesiek2010 replied:
This class no longer exists.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/viewmodels/FormMapViewModel.java:
```diff
@@ -99,6 +102,15 @@ private void initializeFormInstances() {
     }
 
+    public MappableFormInstance getInstanceById(long instanceId) {
+        for (MappableFormInstance mappableFormInstance : mappableFormInstances) {
```
Doing a sequential search here could be slow if there are lots of instances. I might have considered saving the database id to make this lookup fast. It's probably not that important to do now but wanted to flag it for future consideration.

grzesiek2010 replied:
It is removed after removing BottomSheetDialogFragment

lognaturel reviewed (commented):
Looks good! I don't see anything critical to change in this initial release I've made a couple of comments inline and have a couple more considerations below.

Is the graying out of the map new behavior? I think the position is relevant when the bottom sheet is shown so like in Google Maps, I would expect the map to remain active. 

Relatedly, I would expect that when an instance has been tapped and the bottom sheet is showing, tapping on another instance would immediately show me that second instance's information. One use case for this feature is trying to find a record when you know the general area it was captured in. For example, let's say you visited several households and you want to find Bob's household. You might tap a few different instances in the general neighborhood before finding the right one. Having to tap to close the bottom sheet and then again to select another instance adds a bit of friction. Is there an advantage to doing it as it is now?

grzesiek2010 committed:
Removed redundant method

grzesiek2010 committed:
Converted BottomSheetDialogFragment to BottomSheetBehavior

grzesiek2010 commented:
> Is the graying out of the map is new behavior?

> Relatedly, I would expect that when an instance has been tapped and the bottom sheet is showing, tapping on another instance would immediately show me that second instance's information.

I used `BottomSheetDialogFragment` which is modal and works like other DialogFragments that's why...

there are some tricks and workarounds for example to get rid of that dimming and maybe even to make the rest of the screen still enabled (touchable) but nothing seems reliable.
I agree with you that it would be better to make the sheet work in the way you described so that i converted `BottomSheetDialogFragment` to `BottomSheetBehavior` https://github.com/getodk/collect/pull/3778/commits/471a4d8351aba5eb99c3d860ce0efb55b12f65f6

**pros:**
-the map is not dimmed and you can scroll it
-it allows avoiding some strange problems like: [Android BottomSheetDialogFragment does not expand completely](https://stackoverflow.com/questions/37555403/android-bottomsheetdialogfragment-does-not-expand-completely)

**cons:**
-I had to remove the class that extended `BottomSheetDialogFragment` and moved the code to the activity
-the sheet disappears on screen rotation by default (but we can handle it)

so to sum up it works better with `BottomSheetBehavior`  but the code might look worse because I had to remove the separate file that extended `BottomSheetDialogFragment`.

I think we should go with `BottomSheetBehavior` do you agree @lognaturel?

grzesiek2010 committed:
Fixed tests

grzesiek2010 committed:
Added tests

grzesiek2010 committed:
Enable animation when centering map after tapping on a marker

grzesiek2010 committed:
Hide the bottomSheet after opening a form

grzesiek2010 commented:
I can think of a few possible improvements but these might be implemented is separate prs:
1. The marker we clicked on is centered but if there are lots of markers close that one it might be unclear which one we selected so we should highlight it somehow - maybe just bigger marker icon?
2. Clicking the back button closes the FormMapActivity even if the bottom sheet is visible I guess it might be a bit confusing and users might expect that dialog to be closed after clicking the back button.
3. The area below map buttons is transparent and we can see the map but it's not touchable and we can't for example click on a marker that is there.

lognaturel reviewed (approved):
Thanks for making the change to `BottomSheetBehavior`. 👍 I agree that this is great for a first release and that we can follow up shortly with the other enhancements you list.

kkrawczyk123 commented:
**Tested with success!**
Verified on Androids 5.1 and 10.

Tested scenarios:
- light mode
- dark mode
- screen rotation
- minimize app
- extinguish screen
- Google maps
- Mapbox
- OSM
- offline layer 
- all statuses: saved, finalized, sent and failed
- delete answer
- edit answer
- change time and status after edition
- zoom buttons
- add submission through displayed map
- changed language 
- encrypted form

mmarciniak90 commented:
Tested with success on Android: 6.0, 7.0, 8.1, 9.0

@getodk-bot unlabel "needs testing"
@getodk-bot label "behavior verified"
